### DISCUSSION TITLE
Integrate AirSim with other simulators to get a better physics engine for UAVs.

### DISCUSSION THREAD

**Discussion by jonyMarino** [microsoft/AirSim#3284] [Ideas]
AirSim uses its own simplified multirotor physics engine that we call Fast Physics. It is perfect for certain needs but may fall short on others.
Instead of developing Fast Physics further, a solution would be to integrate another simulator such as Gazebo, JSBSim, jMavSim, etc.
We would like to know:
1_ What simulators are you using for physics/dynamics
and what do you think about:
2_ Pros and cons of those simulators regarding its physics engine and the ability to customize for new drones.
3_ The facilities or difficulties to integrate AirSim with each of them (like sending drone state to AirSim and receiving collision info) 

This discussion extends the discussions started in #2955 and #2380.

**Comment by Zartris** [comment]
I like the idea of using Gazebo for the physics.
We have already seen multiple papers utilizing these physic engines and combining them with a much better graphic engine.
[flightmare](https://uzh-rpg.github.io/flightmare/) and [FlightGoogles](https://flightgoggles.mit.edu/) have already shown how a possible solution to integrate gazebo as the backend physics, so a start of implementing this could be like their structure? But this is of course up to debate, and I would love to discuss this further.

**Reply by jonyMarino** [reply]
I will take advantage of the fact that Flightmare is named here and within this context to clear up a mistake that I don't want new people to fall for. Unfortunately, there is an error in the paper they show in that project saying that AirSim uses unreal physics for simulations. As is clear from this discussion, that's only true for the car mode, but we use our own engine for drones, which we call FastPhysics.

**Comment by jonyMarino** [comment]
What I don't like about Gazebo is that it's a runtime integration, with all its performance implications. On the other hand, with a library like JSBSim, we can have a "compilation" integration. I don't know which physical model is more accurate, but I am inclined that JSBSim's should be.
On the other hand, I suppose that many people would want to reuse their work done in Gazebo (Is that so?) I saw the integration made in flightmare, and I don't see the collision info going back to Gazebo. Is that not feasible? 
@Zartris

**Reply by m1baldwin** [reply]
Hey, when we talk about using Gazebo for physics, are we talking really about bringing ODE - the main physics engine for gazebo - into support with Airsim? Or it is another aspect of the physics stack in gazebo we are discussing, like how vehicle models are defined in sdf/urdf files?

Also, I think Gazebo is sort of your de-facto entry level simulator for a lot of robotics development and research these days, so even if many people move on to more “rich” simulation tools, there may be utility in having the ability to still transfer initial work forward to these new tools.

**Reply by jonyMarino** [reply]
Hi Michael!, thanks for joining the discussion!
We are brainstorming here. So anything that is achievable is welcome!
At first, I had thought of communicating with Gazebo through the network to reproduce the movement of the vehicle (and forward the collisions if possible?). But if anyone knows how to use ODE directly, is welcome to contribute here.

**Comment by oogetyboogety** [comment]
Excuse my lack of knowledge on AirSim architecture, how does it use `FastPhysics` as opposed to Unreal Engine's `Chaos`?

**Reply by Zartris** [reply]
Would love to know if we could use the Chaos physics since they are changing to this completely in Unreal Engine 5

**Reply by zimmy87** [reply]
FastPhysics is for drone (quadrotor) physics only, and car physics is done completely by Unreal’s PhysX built-in vehicle model. Currently we are not using Chaos because we are using PhysX instead.

**Reply by oogetyboogety** [reply]
thanks for the explanation. Seems smart. I just saw this https://www.reddit.com/r/unrealengine/comments/kpo9za/chaos_vs_physx_just_tried_new_chaos_physics/?utm_medium=android_app&utm_source=share

**Comment by AOS55** [comment]
Sorry for taking a while to get back on this trying to think more about the general use case and what is useful for development:

For 1, 2 & 3 Around our lab we have quite a range for simulation:

 - Im using JSBSim principally because of how easy it is to implement your flight dynamic model if given the data and it is quite extensible. Being opensource is of course also beneficial and the community working on it is helpful. It is nice how well docuemented the code itself is to and it is a full non-linear model. #2955 documents this challenge already I will update this later on, the python API integration works fine but has problems with collisions. The C++ (native) airsim work I expect will be faster but I have had some challenges debugging the programme. 
 
 - Some people working on high frequency control for RL are using their own FDM they wrote in python but I don't know if they have it freely available/open sourced. Obviously this is very specific though so not really suitable for direct airsim integration. This would be straightforward to add with CV mode. 
 
 - Gazebo is also used by several in the lab, generally due to its integration with ROS I think this is generally constrained to path-planning research where the FDM is less important for our use cases. You seem more knowledgable about Gazebo than myself but I've heard from others that they have found it to model interactions in quite a rigid way, Im not sure if it is a full non-linear 6DOF model. Challenges seem to have already been brought up. 
 
 - Simulink/Matlab is also quite ubiquitious, we have an F-16 FDM based on Stevens & Lewis Aircraft, Control & Simulation textbook that is used for a couple of undergrad courses. Some people also use Simulinks aerospace blockset but I think this is actually JSBSim under the hood. Of course the problem with Matlab is that it is proprietry and would break the open source paradigm which I think is quite important. Also Matlab is generally quite slow as every data-type is a <vector> type. Im guessing Matlab with CV mode would work well if it could just connect to the api server, or you could even run a python intermediary client. This is not ideal though.

I think as has been mentioned before having the means to get collision data out from CV mode would be useful but the way UE4 works with this may be challenging. Ideally given the extensibility of JSBSim if that is added in as its own mode (#2955, what I am trying to do) JSBSim should be sufficient for most non-linear fixed wing flight dynamic models. This should cover a lot of bases as airsim would then have:

- Cars through fastPhysics and PhysX
- Mutirotor through MultirotorPhysics
- FixedWing (helicopters & airships) through JSBSim

My lab is generally focused on Multirotor and fixedwing UAVs but of course automotive automation automation is probably the largest driver and most researched.

**Reply by jonyMarino** [reply]
Thanks, Alexander! Your insights are very valuable for this discussion.

**Reply by jonyMarino** [reply]
Hi @AOS55! I want to add 2 things to what you said.
First, you reference the CV mode several times, but I think a better approach is to use the drone mode and disable the drone's physical update keeping only the position passed through setVehiclePose(); that way, you can get the collisions. You can find an example of that here: https://github.com/jonyMarino/AirSim/tree/external_dynamics_simulator_with_example.
Second, a small correction, fastphysics is the physics model for multirotors, not for cars. 😉

**Reply by AOS55** [reply]
Hi @jonyMarino great thanks for this I've been playing with your branch over the past week or so and it works really well thanks a lot for doing this. Are you planning to merge it back to master too?

I've been working on a python-client app to worki with the FW JSBSim implementation and using your modified external dynamics branch this gives all the functionality I need to get a reward function based on a collision for my RL research. I think it would still be best to eventually do a fixedwing C++ native solution with JSBSim as in #2955 but I have found this to work well for now.

**Reply by AOS55** [reply]
In case anyone wants a quick fixed-wing solution https://github.com/AOS55/Fixedwing-Airsim might be of use and might be a useful prototype of the final architecture of a C++ JSBSim combination. The programme is still in heavy development however and I haven't put together a full quick start guide... yet.

**Reply by jonyMarino** [reply]
Hi Alexander! Thank you very much for the tests.
Yes, the idea is to create a PR with this. This month I should do it.
I am very excited about your contribution. It would be great to have a video and instructions for newbies. Community adoption could speed up compiled C++ integration.

**Reply by jonyMarino** [reply]
Here is the PR #3626. Now you have spinning rotors also.

**Comment by pkashya** [comment]
I'm not a user AirSim yet

But has anyone looked into Chrono (specifically synChrono) physics engines?
https://projectchrono.org/

If we are curious about wheels and not only wings,
Thoughts on CARLA?
https://carla.org/

**Comment by RoninsCode** [comment]
this is may kind of random but, can somebody help me to install airSim with the UE+ and the chaos Engine?
